HTTP2: add error details on some NGHTTP2 errors.

matteof-google opened the pull request:
Add wiring to propagate error details from the http2 codec to the
access log.  Such wiring inserts a DownstreamProtocolError response
flag (appearing as "DPE" in the log) and a string with details of the
error.  Currently, this mechanism is only triggered by the
onInvalidFrame callback.

Before this change, DownstreamProtocolError was handled
inconsistently.  If stream_error_on_invalid_http_messaging, we closed
the stream but did not log a DPE.  Otherwise, we closed all streams
and logged a DPE.  The current code reports a DPE in all cases.

Signed-off-by: Matteo Frigo <matteof@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: add error details on some NGHTTP2 errors
Risk Level: low
Testing: unit tests for specific expected errors, integration tests to verify that the access log has the expected output
Docs Changes: none
Release Notes: none
[Optional Fixes #Issue]
[Optional Deprecated:]

yanavlasov commented on source/common/http/http2/codec_impl.h:
```diff
@@ -230,6 +239,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
     bool pending_receive_buffer_high_watermark_called_ : 1;
     bool pending_send_buffer_high_watermark_called_ : 1;
     bool reset_due_to_messaging_error_ : 1;
+    absl::string_view details_;
```
This assumes that whatever was used to initialize `details_` has static duration. It is right now. However I would sleep better if this was `std::string` and not think about someone calling `setDetails` with parameter of automatic storage duration.

matteof-google replied:
Yan,

I'll be happy to follow your advice, but I want to point out that I copied this pattern from the equivalent http1 code.  Does this change your advice?  (E.g., I could change the http1 code as well.)

yanavlasov replied:
No it does not. I do not think this is a safe pattern.

matteof-google replied:
Thanks.  Fixed in 5e610eab4

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1879,6 +1872,26 @@ void ConnectionManagerImpl::ActiveStream::onResetStream(StreamResetReason, absl:
+  // interpretation appears to be correct in today's code
+  // (2020-02-10).
+  if (reason == StreamResetReason::LocalReset) {
+    stream_info_.setResponseFlag(StreamInfo::ResponseFlag::DownstreamProtocolError);
```
I'm not sure this is correct.
I think in the case there's an error and envoy does sendLocalResponse, maybeEndEncode() will call doEndStream
to indicate that the stream is fully ended (we don't allow continued POST body upload if we're sending a local response) doEndStream will resetStream also with LocalReset. This path can be triggered by other errors, such as unexpected upstream disconnect.

alyssawilk commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -25,6 +25,26 @@ namespace Envoy {
 namespace Http2 {
 
+class Http2ResponseCodeDetailValues {
+  const absl::string_view NgHttp2ErrHttpHeader = "http2.nghttp2_err_http_header";
```
mind looking at the nghttp2 explanations and seeing if we can put something more readable here and/or in comments?
I suspect one is for a particular header being invalid, one is for violating the spec either in the header block or at large.

yanavlasov commented:
@matteof-google please avoid force pushing as it hides previous review comments.

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -393,6 +393,15 @@ void ConnectionManagerImpl::resetAllStreams(
     stream.response_encoder_->getStream().removeCallbacks(stream);
     stream.onResetStream(StreamResetReason::ConnectionTermination, absl::string_view());
     if (response_flag.has_value()) {
+      // TODO(matteof): This code duplicates some of the logic in
```
Are you planning on picking this up later?  If not I'd skip the TODO.

matteof-google replied:
ack, will be in the next push

alyssawilk commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -620,13 +646,18 @@ int ConnectionImpl::onInvalidFrame(int32_t stream_id, int error_code) {
   ENVOY_CONN_LOG(debug, "invalid frame: {} on stream {}", connection_, nghttp2_strerror(error_code),
                  stream_id);
 
+  // Set details of ERROR_CODE in the stream whenever we have one.
```
ERROR_CODE -> error_code

alyssawilk commented on source/common/http/http2/codec_impl.h:
```diff
@@ -176,6 +176,18 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
+    absl::string_view responseDetails() override { return details_; }
+
+    void setDetails(absl::string_view details) {
+      // Figure out whether we can guarantee that details_ is set at
```
Again if you're not going to follow up on this I think you can simply say
// It is probably a mistake to call setDetails() twice, so assert that details_ is empty.

matteof-google replied:
ack, will be in the next push

alyssawilk commented on source/common/http/http2/codec_impl.h:
```diff
@@ -176,6 +176,18 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
+      // twice. For now assert that details_ is empty.
+      ASSERT(details_.empty());
+
+      // make a copy of DETAILS to decouple the lifetime of the two
```
s/make/Make/
s/DETAILS/details

But given at the end of the day these can basically be static strings, can we switch over to the ConstSingleton and change, comment lifetime requirements (or lock down access to where this is set to be non-public) and avoid the string creation?

matteof-google replied:
The strings are already static strings created via ConstSingleton.  However, in comments above, @yanavlasov remarked that this pattern is risky and recommended that I switch to std::string with the implied copy operation.  (I pointed out that the http1 codec has the same issue.)  Are you suggesting that I revert 39a3fa75e ?

alyssawilk commented on test/integration/http2_integration_test.cc:
```diff
@@ -1772,10 +1773,14 @@ TEST_P(Http2FloodMitigationTest, ZerolenHeader) {
   EXPECT_EQ(1,
             test_server_->counter("http.config_test.downstream_cx_delayed_close_timeout")->value());
+  EXPECT_THAT(waitForAccessLog(access_log_name_), HasSubstr("http2.invalid.header.field"));
+  // expect Downstream Protocol Error
```
nit: expect Downstream Protocol Error -> Expect a downstream protocol error.

matteof-google replied:
ack, will be in the next push

mattklein123 commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -25,6 +25,32 @@ namespace Envoy {
+    case NGHTTP2_ERR_HTTP_MESSAGING:
+      return NgHttp2ErrHttpMessaging;
+    default:
+      return NgHttp2ErrUnknown;
```
Do you have coverage for this case?

matteof-google replied:
The two explicit cases are hit in various places in the unit and integration tests.

I don't have a test for the default case, and I am not sure whether it can be triggered.  The theory we discussed with Alyssa was to return unknown, and whoever sees unknown in future logs will dig into it and add another explicit case.

mattklein123 replied:
I think that's fine, but generally there should be some kind of coverage around this, perhaps by making the translation functions static and making sure they do the right thing during targeted unit testing.

matteof-google replied:
@mattklein123 I can add the test that are you asking for, but I see little value in it, since it will look like a duplication of the switch statement, something like:

check(NGTTP2_ERR_..., "foo");
check(NGTTP2_ERR_..., "bar");
check(-1, "quux");

My inclination is to close this as WONTFIX.  If you have other concerns (e.g., you are striving for 100% coverage and this code gets in the way), please let me know and I'll be happy to oblige.

mattklein123 replied:
It doesn't thrill me that we have code that we are not sure whether we can ever hit, and we have no test coverage for it either. I agree though that this type of test is not super useful. Up to you.

mattklein123 commented on source/common/http/http2/codec_impl.h:
```diff
@@ -176,6 +176,17 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
+
+      // make a copy of details to decouple the lifetime of the two
+      // objects details and details_.
+      details_ = static_cast<std::string>(details);
```
For perf reasons it seems like you could just stipulate/assume that the passed strings are constant, but I don't feel strongly about it. Up to you.

matteof-google replied:
Yeah, my original code agreed with you, but yanavlasov requested this change, and now both you and Alyssa seem to agree with my initial code.  I'll talk to Alyssa and decide what to do.

FWIW, the http1 codec assumes that the details_ string is static, so if we decide to be paranoid, I should fix the http1 codec as well.

mattklein123 replied:
> I'll talk to Alyssa and decide what to do.

Alyssa is out this week. My preference us to assume they are static strings, but I agree that if you and @yanavlasov decide otherwise we should be consistent across both codecs here.

matteof-google replied:
I have pushed a patch tentatively rolling back to static strings, while I sort the issue out with @yanavlasov

yanavlasov commented:
Needs DCO fix for some commits.

matteof-google commented:
Thanks @yanavlasov .  I have fixed the DCO issue and squashed the somewhat messy history into one commit.  The patch should be ready for merging now.
